production/ksonnet/promtail/scrape_config.libsonnet

```diff
@@ -40,25 +40,35 @@ config {
         replacement: '$1',
       },
 
-      // But also include the namespace as a separate label, for routing alerts
+      // But also include the namespace, container, pod as separate labels,
+      // for routing alerts and joining with cAdvisor metrics.
       {
         source_labels: ['__meta_kubernetes_namespace'],
         action: 'replace',
         target_label: 'namespace',
       },
-
-      // Rename instances to be the pod name
       {
         source_labels: ['__meta_kubernetes_pod_name'],
         action: 'replace',
-        target_label: 'instance',
+        target_label: 'pod',  // Not 'pod_name', which disappeared in K8s 1.16.
+      },
+      {
+        source_labels: ['__meta_kubernetes_container_name'],
+        action: 'replace',
+        target_label: 'container',  // Not 'container_name', which disappeared in K8s 1.16.
       },
 
-      // Include container_name label
+      // Rename instances to the concatenation of pod:container:port.
+      // All three components are needed to guarantee a unique instance label.
```

Comment:
Can you elaborate on this? I'd expect pod/container would be enough.

Reply:
Hmm… So it is definitely not enough for Prometheus because a single container can expose different metrics ports. I'm not sure if Loki can do something comparable (i.e. multiple "log streams" from a single container). I guess we really want to keep the `instance` labels in sync between Prometheus and Loki, that's kind of the point, but how does Loki deal with a setup where a single container exposes two separate metrics endpoints but only one log stream?

Reply:
I don't think `instance` is necessarily important for consistency, only `namespace`, `pod`, and `container` are. Logs only produce 1 stream (stdout) while a service can expose metrics on multiple ports. These simply don't have a 1:1 relationship of any sort.

Reply:
So how about just not having an `instance` label in Loki? It would make sense semantically. Logs are per container, so `namespace`/`pod`/`container` is sufficient for Loki. Prometheus needs the port in addition, but instead of adding a port label, we use the `instance` label, as it is tradition in Prometheus land, and we even keep it unique on its own to avoid surprises.

To connect Prometheus metrics with Loki logs, you'd go via `namespace`/`pod`/`container`.
Does that make sense? (Also looking at master @tomwilkie .)

Reply:
> So how about just not having an instance label in Loki?

Make sense, as long as namespace/pod/container is consistent.